tilings/algorithms/obstruction_inferral.py

```diff
@@ -0,0 +1,149 @@
+import abc
+from itertools import chain, product
+
+from comb_spec_searcher import InferralRule
+from permuta import Perm, PermSet
+from tilings import Obstruction
+
+
+class ObstructionInferral(abc.ABC):
+    """
+    Algorithm to compute the tiling created by adding all obstructions from
+    `self.potential_new_obs()` that can be added to the tiling.
+    """
+    def __init__(self, tiling):
+        self._tiling = tiling
+
+    @abc.abstractmethod
+    def potential_new_obs(self):
+        """
+        Return an iterable of new obstruction that should be added to the
+        tiling if possible
+        """
+        pass
+
+    def new_obs(self):
+        """
+        Returns the list of new obstruction that can be added to the tiling.
+        """
+        if hasattr(self, '_new_obs'):
+            return self._new_obs
+        newobs = []
+        merged_tiling = self._tiling.merge()
```

Comment:
Avoid merging. See comment on line 47. The is empty function is deterministic. Merging relies on the generation of gridded perms exhaustively, whereas relying on just is empty is a fail-fast approach.

Reply:
Changed it